File: src/Service/ShippingCalculatorService.php
<?php

namespace App\Service;

use App\Entity\Address;

class ShippingCalculatorService
{
    public function calculateForAddress(Address $address, int $itemsCount, float $subtotal): float
    {
        $base = 7.5;

        $region = mb_strtolower((string) ($address->getRegion() ?? ''));
        $regionMultiplier = str_contains($region, 'tunis') ? 1.0 : 1.25;
        $itemsFee = max(0, $itemsCount - 1) * 1.2;
        $valueFee = $subtotal >= 250 ? 0.0 : 2.0;

        return round(($base * $regionMultiplier) + $itemsFee + $valueFee, 2);
    }
}

File: src/main/java/org/example/service/ShippingCalculatorService.java
package org.example.service;

public class ShippingCalculatorService {
    public double calculateForAddress(String region, int itemCount, double subtotal) {
        double base = subtotal >= 120 ? 0.0 : 7.5;

        if (region != null) {
            String r = region.toLowerCase();
            if (r.contains("sud") || r.contains("remote") || r.contains("interieur")) {
                base += 3.0;
            }
        }

        if (itemCount >= 10) {
            base += 2.5;
        }

        return round2(base);
    }

    private double round2(double value) {
        return Math.round(value * 100.0) / 100.0;
    }
}
